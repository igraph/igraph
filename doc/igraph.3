.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\"
.\" Copyright (C) 2006-2021  The igraph development team
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2, or (at your option) any later
.\" version.
.\"
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License with
.\" your Debian GNU/Linux system, in /usr/share/common-licenses/GPL, or with
.\" the dpkg source package as the file COPYING.  If not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH IGRAPH 3 "May 2021" "igraph library"
.SH NAME
igraph \- a library for creating and manipulating graphs
.SH DESCRIPTION
.B igraph
is a C library for complex network analysis and graph theory, with emphasis on
efficiency, portability and ease of use.
.SH DOCUMENTATION
The full documentation can be downloaded from the homepage of the
library:
.RI < https://igraph.org/c/doc >
.SH BUGS
If you think you have found a bug in igraph, feel free to file a bug report
in the issue tracker at:
.RI < https://github.com/igraph/igraph/issues >

.SH AUTHORS
Gabor Csardi <csardi.gabor@gmail.com>,
.br
Tamas Nepusz <ntamas@gmail.com>,
.br
Vincent Traag <vtraag@gmail.com>,
.br
Szabolcs Horvat <szhorvat@gmail.com>,
.br
Fabio Zanini <fabio.zanini@fastmail.fm>,
.br
Daniel Noom <ggatw@outlook.com>
