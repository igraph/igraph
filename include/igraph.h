/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2003-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_H
#define IGRAPH_H

#ifndef _GNU_SOURCE
    #define _GNU_SOURCE 1
#endif

#include "igraph_version.h"
#include "igraph_memory.h"
#include "igraph_error.h"
#include "igraph_random.h"
#include "igraph_progress.h"
#include "igraph_statusbar.h"

#include "igraph_types.h"
#include "igraph_complex.h"
#include "igraph_vector.h"
#include "igraph_matrix.h"
#include "igraph_array.h"
#include "igraph_dqueue.h"
#include "igraph_stack.h"
#include "igraph_heap.h"
#include "igraph_psumtree.h"
#include "igraph_strvector.h"
#include "igraph_vector_ptr.h"
#include "igraph_spmatrix.h"
#include "igraph_sparsemat.h"
#include "igraph_qsort.h"

#include "igraph_constants.h"
#include "igraph_datatype.h"
#include "igraph_iterators.h"
#include "igraph_interface.h"
#include "igraph_constructors.h"
#include "igraph_games.h"
#include "igraph_microscopic_update.h"
#include "igraph_centrality.h"
#include "igraph_paths.h"
#include "igraph_components.h"
#include "igraph_structural.h"
#include "igraph_transitivity.h"
#include "igraph_neighborhood.h"
#include "igraph_topology.h"
#include "igraph_bipartite.h"
#include "igraph_cliques.h"
#include "igraph_layout.h"
#include "igraph_visitor.h"
#include "igraph_community.h"
#include "igraph_conversion.h"
#include "igraph_foreign.h"
#include "igraph_motifs.h"
#include "igraph_operators.h"
#include "igraph_flow.h"
#include "igraph_nongraph.h"
#include "igraph_cocitation.h"
#include "igraph_adjlist.h"
#include "igraph_attributes.h"
#include "igraph_blas.h"
#include "igraph_lapack.h"
#include "igraph_arpack.h"
#include "igraph_mixing.h"
#include "igraph_separators.h"
#include "igraph_cohesive_blocks.h"
#include "igraph_eigen.h"
#include "igraph_hrg.h"
#include "igraph_threading.h"
#include "igraph_interrupt.h"
#include "igraph_scg.h"
#include "igraph_matching.h"
#include "igraph_embedding.h"
#include "igraph_scan.h"
#include "igraph_graphlets.h"
#include "igraph_epidemics.h"
#include "igraph_lsap.h"
#include "igraph_coloring.h"

#endif
