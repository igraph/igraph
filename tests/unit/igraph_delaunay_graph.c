/*
   IGraph library.
   Copyright (C) 2025  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>

#include "test_utilities.h"

igraph_error_t delaunay(
        igraph_real_t *points, igraph_integer_t numpoints,
        igraph_integer_t dimension, igraph_bool_t printing) {

    igraph_matrix_t points_mat;
    igraph_t g;

    IGRAPH_CHECK(igraph_matrix_init_array(&points_mat, points, numpoints, dimension, IGRAPH_ROW_MAJOR));
    IGRAPH_FINALLY(igraph_matrix_destroy, &points_mat);
    IGRAPH_CHECK(igraph_delaunay_graph(&g, &points_mat));
    IGRAPH_FINALLY(igraph_destroy, &g);

    if (printing) {
        printf("%" IGRAPH_PRId "\n", igraph_ecount(&g));
        print_graph_canon(&g);
    }

    igraph_matrix_destroy(&points_mat);
    igraph_destroy(&g);
    IGRAPH_FINALLY_CLEAN(2);
    return IGRAPH_SUCCESS;
}

#define PTCOUNT(points, dim) sizeof(points) / sizeof(points[0]) / dim

int main(void) {
    igraph_real_t points_raw[] = {
        0.311138, 0.59391, 0.67498, 0.1694, 0.822953, 0.904456, 0.129484,
        0.350282, 0.756002, 0.373154, 0.763904, 0.870096, 0.786728,
        0.00985041, 0.406027, 0.504641, 0.174089, 0.358568, 0.388813,
        0.744543, 0.0681765, 0.544161, 0.289917, 0.487095, 0.605402,
        0.270968, 0.978097, 0.955441, 0.495331, 0.688885, 0.693659, 0.579305,
        0.169171, 0.623525, 0.725786, 0.250287, 0.00648143, 0.743068,
        0.71552, 0.534675, 0.757967, 0.533673, 0.804339, 0.113337, 0.445796,
        0.410465, 0.38357, 0.918967, 0.906415, 0.893578, 0.501743, 0.914856,
        0.561324, 0.818718, 0.158764, 0.903373, 0.213128, 0.702836, 0.672937,
        0.291723, 0.855333, 0.0981787, 0.804335, 0.702915, 0.824296,
        0.294826, 0.702266, 0.346332, 0.223656, 0.640166, 0.404121, 0.151721,
        0.32827, 0.290019, 0.875388, 0.221269, 0.260946, 0.443243, 0.97165,
        0.11638, 0.246203, 0.447883, 0.987007, 0.0692745, 0.109563, 0.339652,
        0.0666818, 0.318156, 0.92505, 0.439913, 0.575609, 0.577994, 0.494698,
        0.335231, 0.5605, 0.701834, 0.863751, 0.129853, 0.641124, 0.277378,
        0.897763, 0.429955, 0.331589, 0.925289, 0.980783, 0.835259, 0.291818,
        0.693408, 0.631442, 0.53206, 0.159631, 0.350867, 0.699043, 0.993084,
        0.638868, 0.933556, 0.524775, 0.598018, 0.565078, 0.639017, 0.480785,
        0.0698051, 0.60897, 0.350966, 0.411467, 0.0688254, 0.430729,
        0.761323, 0.359096, 0.590563, 0.372144, 0.1754, 0.422717, 0.296634,
        0.874941, 0.32733, 0.616607, 0.547951, 0.902168, 0.369618, 0.226083,
        0.10907, 0.0655338, 0.456096, 0.633249, 0.676374, 0.9357, 0.945376,
        0.581192, 0.95893, 0.655664, 0.316474, 0.752296, 0.14192, 0.0125042,
        0.595269, 0.084296, 0.281774, 0.6996, 0.66766, 0.478567, 0.505426,
        0.716551, 0.232743, 0.191, 0.0114542, 0.609768, 0.903796, 0.772037,
        0.60114, 0.758929, 0.770639, 0.968881, 0.382973, 0.248955, 0.290848,
        0.752831, 0.0109218, 0.903305, 0.762493, 0.696122, 0.91684,
        0.0138796, 0.45168, 0.668332, 0.0793168, 0.499606, 0.321686,
        0.0934459, 0.499221, 0.92654, 0.917383, 0.479325, 0.147302, 0.9623,
        0.124589, 0.752746, 0.127419, 0.211051, 0.410824
    };

    igraph_real_t point_cloud_3d[] = {
        0.125269, 0.195444, 0.285089, 0.992534, 0.250088, 0.551803, 0.104234,
        0.99641, 0.72404, 0.00427073, 0.721804, 0.00514948, 0.0685917,
        0.504841, 0.884932, 0.320637, 0.22425, 0.107845, 0.460224, 0.923877,
        0.480411, 0.139484, 0.687409, 0.871486, 0.0273725, 0.518508,
        0.472275, 0.57603, 0.848422, 0.79629, 0.844017, 0.51724, 0.486733,
        0.886948, 0.41991, 0.626941, 0.575853, 0.399617, 0.415894, 0.478654,
        0.586383, 0.661319, 0.413907, 0.836301, 0.893907, 0.887531, 0.451777,
        0.392684, 0.488796, 0.1865, 0.518422, 0.087241, 0.424709, 0.210399,
        0.930499, 0.503204, 0.44653, 0.0644197, 0.375475, 0.510661, 0.931541,
        0.742055, 0.739896, 0.320497, 0.949405, 0.00282601, 0.160896,
        0.712019, 0.966189, 0.172367, 0.216689, 0.132008, 0.955499, 0.276767,
        0.271158, 0.085173, 0.306008, 0.621809, 0.826937, 0.702707, 0.367152,
        0.0914104, 0.315055, 0.726888, 0.759314, 0.759621, 0.0954736,
        0.35047, 0.322236, 0.676948, 0.70161, 0.967861, 0.80878, 0.345871,
        0.617758, 0.0890548, 0.947943, 0.93674, 0.00637263, 0.798722, 0.1123,
        0.883446, 0.270618, 0.859522, 0.527832, 0.504072, 0.561741, 0.614159,
        0.611909, 0.779264, 0.0651863, 0.760387, 0.934354, 0.00447453,
        0.652217, 0.00252835, 0.13965, 0.281917, 0.408229, 0.140262,
        0.951417, 0.945453, 0.8125, 0.404407, 0.457046, 0.839765, 0.695236,
        0.345371, 0.372619, 0.33457, 0.484926, 0.329866, 0.813254, 0.0181675,
        0.294805, 0.616989, 0.403054, 0.123351, 0.137132, 0.641034, 0.804549,
        0.776808, 0.193548, 0.0441299, 0.823612, 0.919071, 0.559754,
        0.636797, 0.0478928, 0.36961
    };

    igraph_real_t point_cloud_4d[] = {
        0.768464, 0.05359, 0.69102, 0.246049, 0.996004, 0.189829, 0.00302669,
        0.187796, 0.44531, 0.711051, 0.946157, 0.764919, 0.583837, 0.308961,
        0.219333, 0.885349, 0.480686, 0.620953, 0.0128075, 0.650813,
        0.717234, 0.157298, 0.775933, 0.816187, 0.189236, 0.194153, 0.347867,
        0.942655, 0.177974, 0.120843, 0.298, 0.779916, 0.00154216, 0.755666,
        0.537359, 0.00281348, 0.0687358, 0.298396, 0.19734, 0.940497
    };

    igraph_real_t points_lattice[] = {
        0,0,
        0,1,
        0,2,
        1,0,
        1,1,
        1,2,
        2,0,
        2,1,
        2,2
    };

    igraph_real_t points_collinear[] = {
        0,0,
        2,0,
        0,2,
        1,0
    };

    igraph_real_t points_not_quite_collinear[] = {
        0,0,
        2,0,
        0,2,
        1,0.0000000001
    };

    igraph_real_t points_all_collinear[] = {
        1,1,
        2,2,
        3,3,
        4,4,
    };

    igraph_real_t points_all_coplanar[] = {
            0., 0., 0.,
            -0.9852739637886722, 0.00098610486426615, 0.17098024411421095,
            0.14677925443983744, 0.5177782140612408, 0.842829503226861,
            -0.8384947093488347, 0.518764318925507, 1.0138097473410719
    };

    igraph_real_t points_duplicate[] ={
        1,1,
        0,1,
        1,0,
        1,1
    };

    printf("100 point cloud in 2d\n");
    delaunay(points_raw, PTCOUNT(points_raw, 2), 2, true);

    printf("50 point cloud in 3d\n");
    delaunay(point_cloud_3d, PTCOUNT(point_cloud_3d, 3), 3, true);

    printf("10 point cloud in 4d\n");
    delaunay(point_cloud_4d, PTCOUNT(point_cloud_4d, 4), 4, true);

    printf("3x3 square lattice\n");
    delaunay(points_lattice, PTCOUNT(points_lattice, 2), 2, true);

    printf("Triangle with one subdivided edge\n");
    delaunay(points_collinear, PTCOUNT(points_collinear, 2), 2, true);

    printf("Triangle with one subdivided edge, point then slightly perturbed\n");
    delaunay(points_not_quite_collinear, PTCOUNT(points_not_quite_collinear, 2), 2, true);

    /* TODO: Planned to be supported without error in the future. */
    printf("4 points on a line in 2d\n");
    CHECK_ERROR(delaunay(points_all_collinear, PTCOUNT(points_all_collinear, 2), 2, true), IGRAPH_EINVAL);

    /* TODO: Planned to be supported without error in the future. */
    printf("4 points on a plane in 3d\n");
    CHECK_ERROR(delaunay(points_all_coplanar, PTCOUNT(points_all_coplanar, 3), 3, true), IGRAPH_EINVAL);

    printf("Identical points\n");
    CHECK_ERROR(delaunay(points_duplicate, PTCOUNT(points_duplicate, 2), 2,false), IGRAPH_EINVAL);

    /* The below coordinate arrays are used only partially, hence no PTCOUNT */

    printf("One point in 2d\n");
    delaunay(points_raw, 1, 2, true);

    printf("No points in 2d\n");
    delaunay(points_raw, 0, 2, true);

    printf("0x0 matrix\n");
    delaunay(points_raw, 0, 0, true);

    printf("2 points in 0d\n");
    CHECK_ERROR(delaunay(points_raw, 2, 0, true), IGRAPH_EINVAL);

    printf("4 points in 3d\n");
    delaunay(point_cloud_3d, 4, 3, true);

    /* TODO: Planned to be supported without error in the future. */
    printf("3 points in 3d\n");
    CHECK_ERROR(delaunay(point_cloud_3d, 3, 3, true), IGRAPH_EINVAL);

    VERIFY_FINALLY_STACK();

    return 0;
}
