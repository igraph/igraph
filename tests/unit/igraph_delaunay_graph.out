100 point cloud in 2d
287
directed: false
vcount: 100
edges: {
0 11
0 34
0 53
0 64
1 12
1 49
1 76
1 81
1 92
1 96
1 98
2 5
2 24
2 56
2 73
2 89
2 90
2 95
3 8
3 42
3 55
3 71
3 99
4 17
4 19
4 20
4 32
4 33
4 50
4 67
4 69
5 85
5 89
5 90
6 21
6 30
6 41
6 82
6 88
7 11
7 22
7 64
7 80
8 55
8 87
8 99
9 14
9 23
9 51
9 53
9 63
9 64
10 16
10 71
10 77
10 91
10 94
11 16
11 22
11 34
11 38
11 40
11 64
12 49
12 61
12 75
12 93
12 96
13 52
13 56
13 73
13 95
14 26
14 47
14 58
14 59
14 63
14 64
15 19
15 54
15 68
15 72
15 79
15 84
16 18
16 28
16 34
16 40
16 77
16 94
17 29
17 32
17 33
17 37
17 76
17 81
18 27
18 28
18 77
18 91
19 20
19 33
19 54
19 84
20 44
20 50
20 84
21 30
21 37
21 48
21 76
21 88
21 98
22 36
22 38
22 46
22 61
22 66
22 80
23 25
23 51
23 63
23 74
24 52
24 89
24 95
25 26
25 63
25 74
25 83
26 47
26 63
26 72
26 83
26 85
27 28
27 51
27 56
28 34
28 51
28 53
29 33
29 49
29 75
29 81
30 41
30 48
30 97
31 79
31 84
31 85
31 89
32 37
32 67
33 54
33 61
33 75
34 53
35 62
35 65
35 66
35 96
36 38
36 65
36 66
36 87
37 48
37 67
37 76
37 86
37 97
38 40
38 87
38 99
39 41
39 86
39 97
40 94
40 99
41 52
41 86
41 97
42 43
42 55
42 71
42 78
43 71
43 78
43 82
43 91
44 50
44 52
44 69
44 84
44 86
44 89
45 58
45 59
45 68
45 72
45 80
46 61
46 66
46 93
47 59
47 72
48 97
49 75
49 81
50 69
51 53
51 56
51 74
52 86
52 89
52 95
53 64
54 61
54 68
54 80
55 78
55 87
56 57
56 73
56 74
56 90
57 74
57 83
57 90
58 59
58 64
58 80
59 72
60 62
60 82
60 88
60 92
60 96
61 75
61 80
61 93
62 65
62 70
62 82
62 96
64 80
65 66
65 70
65 87
66 93
66 96
67 69
67 86
68 72
68 80
69 86
70 78
70 82
70 87
71 91
71 94
71 99
72 79
72 85
73 95
74 83
76 81
76 98
77 91
78 82
78 87
79 84
79 85
82 88
83 85
83 90
84 89
85 89
85 90
86 97
87 99
88 92
88 98
92 96
92 98
93 96
94 99
}
50 point cloud in 3d
289
directed: false
vcount: 50
edges: {
0 3
0 5
0 8
0 16
0 17
0 19
0 23
0 25
0 27
0 29
0 33
0 38
0 39
0 43
0 49
1 11
1 15
1 16
1 18
1 20
1 24
1 32
1 33
1 40
1 42
1 44
1 49
2 3
2 4
2 6
2 7
2 8
2 9
2 14
2 21
2 22
2 30
2 32
2 34
2 37
2 46
3 5
3 8
3 17
3 21
3 23
3 31
3 32
3 34
3 37
3 39
3 47
4 7
4 8
4 22
4 27
4 33
4 41
4 46
5 12
5 16
5 23
5 31
5 38
5 39
5 43
5 45
5 47
5 49
6 9
6 13
6 14
6 21
6 26
6 28
6 30
6 31
6 34
6 35
6 36
6 37
6 43
6 48
7 14
7 22
7 34
7 41
7 46
8 13
8 17
8 19
8 25
8 27
8 29
8 31
8 34
8 43
8 46
9 13
9 14
9 20
9 30
9 34
9 35
9 40
9 41
9 48
10 11
10 12
10 15
10 18
10 20
10 26
10 35
10 42
10 48
11 12
11 13
11 15
11 16
11 18
11 20
11 29
11 33
11 35
11 41
11 42
12 16
12 26
12 29
12 35
12 36
12 42
12 43
12 45
12 49
13 14
13 20
13 29
13 34
13 35
13 41
13 43
13 46
14 20
14 22
14 30
14 33
14 34
14 40
14 41
14 46
15 18
15 24
15 26
15 28
15 32
15 42
15 45
16 19
16 25
16 27
16 29
16 33
16 35
16 41
16 42
16 43
16 49
17 19
17 23
17 31
17 39
17 43
18 20
18 24
18 26
18 32
18 40
19 25
19 27
19 29
19 43
19 46
20 26
20 30
20 32
20 33
20 35
20 40
20 41
20 48
21 30
21 31
21 32
21 34
21 36
21 37
21 47
22 33
22 40
22 41
23 38
23 39
23 47
24 28
24 32
24 42
24 44
24 45
24 47
25 27
25 29
25 46
26 28
26 32
26 35
26 36
26 37
26 42
26 43
26 45
26 48
27 29
27 33
27 41
27 46
27 49
28 32
28 36
28 37
28 45
28 47
28 48
29 33
29 35
29 41
29 43
29 46
30 32
30 37
30 40
30 48
31 34
31 36
31 39
31 43
31 45
31 47
32 37
32 40
32 47
32 48
33 38
33 40
33 41
33 44
33 49
34 35
34 43
34 46
35 41
35 43
35 48
36 37
36 43
36 45
36 47
37 47
37 48
38 42
38 44
38 45
38 47
38 49
39 43
39 45
40 48
41 46
42 44
42 45
42 47
42 49
43 45
44 47
44 49
45 47
45 49
47 49
}
10 point cloud in 4d
39
directed: false
vcount: 10
edges: {
0 1
0 2
0 3
0 4
0 5
0 7
0 8
1 2
1 3
1 4
1 5
1 7
1 8
1 9
2 3
2 4
2 5
2 6
2 7
2 8
2 9
3 4
3 5
3 6
3 7
3 9
4 5
4 6
4 7
4 8
4 9
5 6
5 7
6 7
6 8
6 9
7 8
7 9
8 9
}
3x3 square lattice
16
directed: false
vcount: 9
edges: {
0 1
0 3
1 2
1 3
1 4
1 5
2 5
3 4
3 6
3 7
4 5
4 7
5 7
5 8
6 7
7 8
}
Triangle with one subdivided edge
5
directed: false
vcount: 4
edges: {
0 2
0 3
1 2
1 3
2 3
}
Triangle with one subdivided edge, point then slightly perturbed
6
directed: false
vcount: 4
edges: {
0 1
0 2
0 3
1 2
1 3
2 3
}
4 points on a line in 2d
4 points on a plane in 3d
Identical points
One point in 2d
0
directed: false
vcount: 1
edges: {
}
No points in 2d
0
directed: false
vcount: 0
edges: {
}
0x0 matrix
0
directed: false
vcount: 0
edges: {
}
2 points in 0d
4 points in 3d
6
directed: false
vcount: 4
edges: {
0 1
0 2
0 3
1 2
1 3
2 3
}
3 points in 3d
